File: ops/workers/src/evv_workers/probe.py
from __future__ import annotations

import json
import subprocess
from dataclasses import dataclass


class ProbeError(RuntimeError):
    pass


@dataclass(frozen=True)
class Probe:
    duration_s: float
    width: int | None
    height: int | None
    video_codec: str | None
    audio_codec: str | None
    pix_fmt: str | None
    sample_rate: int | None
    profile: str | None


def ffprobe(path: str) -> Probe:
    proc = subprocess.run(
        [
            "ffprobe",
            "-v",
            "error",
            "-print_format",
            "json",
            "-show_streams",
            "-show_format",
            path,
        ],
        capture_output=True,
        text=True,
    )
    if proc.returncode != 0:
        raise ProbeError(proc.stderr.strip() or "ffprobe failed")
    payload = json.loads(proc.stdout)
    streams = payload.get("streams")
    if not isinstance(streams, list):
        streams = []
    video = next((s for s in streams if isinstance(s, dict) and s.get("codec_type") == "video"), None)
    audio = next((s for s in streams if isinstance(s, dict) and s.get("codec_type") == "audio"), None)
    fmt = payload.get("format") if isinstance(payload.get("format"), dict) else {}
    duration_raw = fmt.get("duration") if isinstance(fmt, dict) else None
    duration = float(duration_raw) if duration_raw not in (None, "N/A") else 0.0
    return Probe(
        duration_s=duration,
        width=int(video["width"]) if isinstance(video, dict) and video.get("width") else None,
        height=int(video["height"]) if isinstance(video, dict) and video.get("height") else None,
        video_codec=str(video["codec_name"]) if isinstance(video, dict) and video.get("codec_name") else None,
        audio_codec=str(audio["codec_name"]) if isinstance(audio, dict) and audio.get("codec_name") else None,
        pix_fmt=str(video["pix_fmt"]) if isinstance(video, dict) and video.get("pix_fmt") else None,
        sample_rate=int(audio["sample_rate"]) if isinstance(audio, dict) and audio.get("sample_rate") else None,
        profile=str(video["profile"]) if isinstance(video, dict) and video.get("profile") else None,
    )


def require_video(probe: Probe) -> Probe:
    if probe.video_codec is None or probe.duration_s <= 0:
        raise ProbeError("reject: audio-only or non-video")
    return probe

File: ops/apps/app/src/lib/probe.ts
import { spawn } from "node:child_process";
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import path from "node:path";

import { UploadError, type ProbeResult } from "./types";

interface FfprobeStream {
  codec_type?: string;
  codec_name?: string;
  width?: number;
  height?: number;
}

interface FfprobePayload {
  streams?: FfprobeStream[];
  format?: { duration?: string };
}

function run(command: string, args: string[]): Promise<{ code: number; stdout: string; stderr: string }> {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, { stdio: ["ignore", "pipe", "pipe"] });
    const out: Buffer[] = [];
    const err: Buffer[] = [];
    child.stdout.on("data", (chunk: Buffer) => out.push(chunk));
    child.stderr.on("data", (chunk: Buffer) => err.push(chunk));
    child.on("error", reject);
    child.on("close", (code) => {
      resolve({
        code: code ?? 1,
        stdout: Buffer.concat(out).toString("utf8"),
        stderr: Buffer.concat(err).toString("utf8"),
      });
    });
  });
}

export async function probeBytes(bytes: Uint8Array): Promise<ProbeResult> {
  const dir = await mkdtemp(path.join(tmpdir(), "evv-probe-"));
  const file = path.join(dir, "source.bin");
  try {
    await writeFile(file, bytes);
    const proc = await run(process.env.FFPROBE_PATH ?? "ffprobe", [
      "-v",
      "error",
      "-print_format",
      "json",
      "-show_streams",
      "-show_format",
      file,
    ]);
    if (proc.code !== 0) {
      throw new UploadError("reject: audio-only or non-video", 400);
    }
    const payload = JSON.parse(proc.stdout) as FfprobePayload;
    const streams = Array.isArray(payload.streams) ? payload.streams : [];
    const video = streams.find((stream) => stream.codec_type === "video");
    const durationRaw = payload.format?.duration;
    const durationS = durationRaw && durationRaw !== "N/A" ? Number(durationRaw) : 0;
    if (!video?.codec_name || !(durationS > 0)) {
      throw new UploadError("reject: audio-only or non-video", 400);
    }
    return {
      durationS,
      width: typeof video.width === "number" ? video.width : null,
      height: typeof video.height === "number" ? video.height : null,
      videoCodec: video.codec_name,
    };
  } finally {
    await rm(dir, { recursive: true, force: true });
  }
}
